criu/seize.c

```diff
@@ -599,6 +599,34 @@ static int freeze_processes(void)
 				goto err;
 			}
 			nanosleep(&req, NULL);
+
+			if (cgroup_v2)
+				continue;
+
+			/* As per older kernel docs (freezer-subsystem.txt before
+			 * the kernel commit ef9fe980c6fcc1821), if FREEZING is seen,
+			 * userspace should either retry or thaw. While current
+			 * kernel cgroup v1 docs no longer mention a need to retry,
+			 * even recent kernels can't reliably freeze a cgroup v1.
+			 *
+			 * Let's keep asking the kernel to freeze from time to time.
+			 * In addition, do occasional thaw/sleep/freeze.
+			 *
+			 * This is still a game of chances (the real fix belongs to the kernel)
+			 * but these kludges might improve the probability of success.
+			 *
+			 * Cgroup v2 does not have this problem.
+			 */
+			switch (i%32) {
+				case 30:
+					freezer_write_state(fd, THAWED);
```

Comment:
I think it is better to do that before nanosleep, otherwise we waste one time slice.

Reply:
We don't waste anything, we just spend one time slice in THAWED state, letting the processes to run for some time, hoping this will change their state to one in which they will have a better chance to freeze.

The only "wasteful" check is `get_freezer_state` after this, since it will return `THAWED`, and I don't consider it heavy enough to worth optimizing out and making the code more complicated than it should be.

Or, the code can be changed like this, to be more straightforward in what we're trying to do here:

```diff
@@ -620,10 +621,11 @@ static int freeze_processes(void)
                        switch (i%32) {
                                case 30:
                                        freezer_write_state(fd, THAWED);
+                                       nanosleep(&req, NULL);
+                                       freezer_write_state(fd, FROZEN);
                                        break;
                                case 9:
                                case 20:
-                               case 31:
                                        freezer_write_state(fd, FROZEN);
                                        break;
                        }
```

Reply:
```
			nanosleep(&req, NULL); <---- we should not sleep here if we are going to thaw the cgroup.

			if (cgroup_v2)
				continue;

			/* As per older kernel docs (freezer-subsystem.txt before
```

Reply:
Yes, my bad. Moved nanosleep to before checking the status -- the difference is the minimum sleep time is 100ms now (was 0). Not sure if it makes any practical difference. If it does, we can do
```c
if i != 0
   nanosleep(&req, NULL);
```

Reply:
You can do 1M syscalls for 100ms. We probably need to reconsider this timeout. In many cases, the full container dump should take a shorter period.

let's do something like that:
```
+static void cgroupv1_freezer_kludges(int fd, int i, const struct timespec *req)
+{
+       /* As per older kernel docs (freezer-subsystem.txt before
+        * the kernel commit ef9fe980c6fcc1821), if FREEZING is seen,
+        * userspace should either retry or thaw. While current
+        * kernel cgroup v1 docs no longer mention a need to retry,
+        * even recent kernels can't reliably freeze a cgroup v1.
+        *
+        * Let's keep asking the kernel to freeze from time to time.
+        * In addition, do occasional thaw/sleep/freeze.
+        *
+        * This is still a game of chances (the real fix belongs to the kernel)
+        * but these kludges might improve the probability of success.
+        *
+        * Cgroup v2 does not have this problem.
+        */
+       switch (i % 32) {
+       case 9:
+       case 20:
+               freezer_write_state(fd, FROZEN);
+               break;
+       case 31:
+               freezer_write_state(fd, THAWED);
+               nanosleep(req, NULL);
+               freezer_write_state(fd, FROZEN);
+               break;
+       }
+}
+
 static int freeze_processes(void)
 {
        int fd, exit_code = -1;
        enum freezer_state state = THAWED;
 
        static const unsigned long step_ms = 100;
+       /* Since opts.timeout is in seconds, multiply it by 1000 to convert to milliseconds. */
        unsigned long nr_attempts = (opts.timeout * 1000) / step_ms;
        unsigned long i = 0;
 
@@ -598,6 +628,10 @@ static int freeze_processes(void)
                                pr_err("Unable to freeze cgroup %s (timed out)\n", opts.freeze_cgroup);
                                goto err;
                        }
+
+                       if (!cgroup_v2)
+                               cgroupv1_freezer_kludges(fd, i, &req);
+
                        nanosleep(&req, NULL);
               
```

Reply:
Makes sense, although in this case we have a wasteful nanosleep when `i` reaches `nr_attempts` and the timeout is not hit. I.e. we check the state, sleep, and then print an error. Not a new problem, but let me fix this, too.